tapdb/multiverse.go

```diff
@@ -61,21 +70,238 @@ type BatchedMultiverse interface {
 	BatchedTx[BaseMultiverseStore]
 }
 
+// ProofKey is used to uniquely identify a proof within a universe. This is
+// used for the LRU cache for the proofs themselves, which are considered to be
+// immutable.
+type ProofKey [32]byte
+
+// NewProofKey takes a universe identifier and leaf key, and returns a proof
+// key.
+func NewProofKey(id universe.Identifier, key universe.LeafKey) ProofKey {
+	idBytes := id.Bytes()
+	leafKeyBytes := key.UniverseKey()
+
+	// The proof key maps down the ID and the leaf key into a single
+	// 32-byte value: sha256(id || leaf_key)..
+	h := sha256.New()
+	h.Write(idBytes[:])
+	h.Write(leafKeyBytes[:])
+
+	return fn.ToArray[ProofKey](h.Sum(nil))
+}
+
+// numCachedProofs is the number of universe proofs we'll cache.
+const numCachedProofs = 25_000
+
+// cachedProof is a single cached proof.
+type cachedProof []*universe.Proof
+
+// Size just returns 1 as we're limiting based on the total number of proofs.
```

Comment:
Related to the comment above this would the number of groups in the cache but the actual number of proofs can well exceed 25k if there are a lot of large groups.

Reply:
So this actually just always maps to one value, the API on the DB level lets you try to get stuff like all the proofs for a given outpoint (not binding script key). We go `treeID -> proofKey`. Each of the new group instances will have a diff `proofKey`.

Reply:
Do you mean that `FetchProofLeaf` is always called with the `universeKey.ScriptKey` set, so the returned array is always one element?

If so we could assert that before updating the cache and then make `cachedProof` not an array - was also confused by this and the impact on cache size.

Reply:
> Do you mean that FetchProofLeaf is always called with the universeKey.ScriptKey set, so the returned array is always one element?


Yeah it's always a single element. The functionality to expose the partial queries isn't exposed over RPC. 

The current limits were set just to make sure we could cache everything as is to stop hammering the disk. W/ the current "sync all behavior" and some of the problematic clients, we'd end up caching everything as the LRU size would quickly be exceeded. Agree we can tune some of these values down some though. At present, there're ~42k total assets.